modules/optimizing-by-encapsulation.adoc

```diff
@@ -0,0 +1,38 @@
+// Module included in the following assemblies:
+//
+// * scalability_and_performance/optimizing-cpu-usage
+
+[id="optimizing-cpu-usage_{context}"]
+= Encapsulating mount namespaces
+
+Mount namespaces are used to isolate mount points so that processes in different namespaces cannot view each others' files. Encapsulation is the process of moving Kubernetes mount namespaces to an alternative location where they will not be constantly scanned by the host operating system.
+
+The host operating system uses systemd to constantly scan all mount namespaces: both the standard Linux mounts and the numerous mounts that Kubernetes uses to operate. The current implementation of Kubelet and CRI-O both use the top-level namespace for all container runtime and Kubelet mountpoints. However, encapsulating these container-specific mountpoints in a private namespace reduces systemd overhead with no difference in functionality. Using a separate mount namespace for both CRI-O and Kubelet can encapsulate container-specific mounts from any systemd or other host OS interaction.
+
+This ability to potentially achieve major CPU optimization is now available to all {product-title} administrators. Encapsulation can also improve security, by storing Kubernetes-specific mount points in a location safe from inspection by unprivileged users.
+
+The following diagrams illustrate a Kubernetes installation before and after encapsulation. Both scenarios show example containers which have mount propagation settings of bidirectional, host-to-container, and none.
+
+image::before-k8s-mount-propagation.png[Before encapsulation]
+
+Here we see systemd, host OS processes, Kubelet, and the container runtime sharing a single mount namespace.
+
+* systemd, host OS processes, Kubelet, and the container runtime each have access to and visibility of all mount points.
+
+* A mount originating in Container 1, such as `/run/a` is visible to systemd, host OS processes, Kubelet, container runtime, and other containers with host-to-container or bidirectional mount propagation configured.
```

Comment:
In the "after" part, we have a bullet about Container2 and Container3 as well - but they would be identical to the text in the "after" section.

Reply:
Unclear about required edit?

Reply:
Do we want the "before" and "after" sections to each have a note about each container, or only when the behavior is different?  It seems odd to me that this first section only has a note about Container 1 (which is my fault, I considered the 2 sections separately at first).

One idea might be to move the points from Container2 and Container3 from the "after" section to the "before" section, and then add a short note for each in the "after" section to just say something like "The behavior of Container 2 and Container 3 is unchanged"...

Reply:
OK will consider what to do.

Reply:
Done.